tests/unit/test_sentry_client.py

```diff
@@ -283,3 +284,170 @@ def test_upload_installable_app_single_chunk_failure(self):
         # Each request can itself be retried 3 times
         # So (1*3)*4 = 16
         assert upload.call_count == 16
+
+    @responses.activate
+    def test_download_artifact_success(self):
+        """Test successful download of artifact."""
+        # Mock HEAD request for file size
+        responses.add(
+            responses.HEAD,
+            "https://example.com/api/0/internal/test-org/test-project/files/preprodartifacts/test-artifact/",
+            headers={"Content-Length": "40"},  # 40 bytes total
+        )
+
+        # Mock GET request for download
+        responses.add(
+            responses.GET,
+            "https://example.com/api/0/internal/test-org/test-project/files/preprodartifacts/test-artifact/",
+            body=b"A" * 20 + b"B" * 20,  # 40 bytes total, will be 2 chunks of 20MB each (but truncated)
+        )
+
+        client = SentryClient(base_url="https://example.com", shared_secret="password")
+        out = io.BytesIO()
+
+        result = client.download_artifact("test-org", "test-project", "test-artifact", out)
+
+        assert result == 40
+        out.seek(0)
+        assert out.read() == b"A" * 20 + b"B" * 20
+
+    @responses.activate
+    def test_download_artifact_no_head_request(self):
+        """Test download when HEAD request fails."""
```

Comment:
In what case could the HEAD request fail but the GET request succeed? If the HEAD request fails couldn't we just mark the entire thing as failed?

Reply:
the HEAD request is _technically_ optional, as it's only used to get the Content-Length header for progress reporting

Frankly the only valid reason where HEAD fails but GET succeeds would be due to some transient network issue. Otherwise I expect them to both succeed, it just seemed better to allow HEAD to be optional

Reply:
It's technically optional in a vacuum, but since we are the ones controlling the business logic can't we make it required? 😆 I'm just a bit confused why we are supporting so many different configurations unless there's something I'm missing.